File: apps/meal-planner/src/types/recipe.ts
export interface IngredientInRecipe {
  id?: number
  ingredient_id?: number
  name: string
  quantity?: string
  unit?: string
  notes?: string
  sort_order: number
}

export interface RecipeListItem {
  id: number
  title: string
  category?: string
  cuisine?: string
  image_filename?: string
  servings: number
  prep_time_mins?: number
  cook_time_mins?: number
  is_custom: boolean
  tags?: string
}

export interface Recipe extends RecipeListItem {
  description?: string
  instructions: string
  source_url?: string
  external_id?: string
  calories?: number
  protein_g?: number
  carbs_g?: number
  fat_g?: number
  created_at: string
  updated_at: string
  ingredients: IngredientInRecipe[]
}

export interface PaginatedRecipes {
  items: RecipeListItem[]
  total: number
  page: number
  per_page: number
  pages: number
}

export interface RecipeFilters {
  q?: string
  ingredient?: string
  category?: string
  cuisine?: string
  is_custom?: boolean
  page?: number
  per_page?: number
}

export interface RecipeFormData {
  title: string
  description: string
  instructions: string
  category: string
  cuisine: string
  servings: number
  prep_time_mins: string
  cook_time_mins: string
  image_filename: string
  source_url: string
  tags: string
  calories: number | ''
  protein_g: number | ''
  carbs_g: number | ''
  fat_g: number | ''
  ingredients: IngredientInRecipe[]
}

File: server/app/schemas/recipe.py
from __future__ import annotations
from pydantic import BaseModel, field_validator
from datetime import datetime
from typing import Optional


class IngredientBase(BaseModel):
    name: str
    category: Optional[str] = None


class IngredientResponse(IngredientBase):
    id: int

    model_config = {"from_attributes": True}


class IngredientInRecipe(BaseModel):
    ingredient_id: Optional[int] = None  # None = look up / create by name
    name: str
    quantity: Optional[str] = None
    unit: Optional[str] = None
    notes: Optional[str] = None
    sort_order: int = 0


class IngredientInRecipeResponse(IngredientInRecipe):
    id: int
    ingredient_id: int

    model_config = {"from_attributes": True}


class RecipeCreate(BaseModel):
    title: str
    description: Optional[str] = None
    instructions: str = ""
    category: Optional[str] = None
    cuisine: Optional[str] = None
    servings: int = 4
    prep_time_mins: Optional[int] = None
    cook_time_mins: Optional[int] = None
    image_filename: Optional[str] = None
    source_url: Optional[str] = None
    tags: Optional[str] = None
    calories: Optional[float] = None
    protein_g: Optional[float] = None
    carbs_g: Optional[float] = None
    fat_g: Optional[float] = None
    ingredients: list[IngredientInRecipe] = []

    @field_validator("title")
    @classmethod
    def title_not_empty(cls, v: str) -> str:
        if not v.strip():
            raise ValueError("title cannot be empty")
        return v.strip()


class RecipeUpdate(BaseModel):
    title: Optional[str] = None
    description: Optional[str] = None
    instructions: Optional[str] = None
    category: Optional[str] = None
    cuisine: Optional[str] = None
    servings: Optional[int] = None
    prep_time_mins: Optional[int] = None
    cook_time_mins: Optional[int] = None
    image_filename: Optional[str] = None
    source_url: Optional[str] = None
    tags: Optional[str] = None
    calories: Optional[float] = None
    protein_g: Optional[float] = None
    carbs_g: Optional[float] = None
    fat_g: Optional[float] = None
    ingredients: Optional[list[IngredientInRecipe]] = None


class RecipeListItem(BaseModel):
    id: int
    title: str
    category: Optional[str] = None
    cuisine: Optional[str] = None
    image_filename: Optional[str] = None
    servings: int
    prep_time_mins: Optional[int] = None
    cook_time_mins: Optional[int] = None
    is_custom: bool
    tags: Optional[str] = None

    model_config = {"from_attributes": True}


class RecipeResponse(BaseModel):
    id: int
    title: str
    description: Optional[str] = None
    instructions: str
    category: Optional[str] = None
    cuisine: Optional[str] = None
    servings: int
    prep_time_mins: Optional[int] = None
    cook_time_mins: Optional[int] = None
    image_filename: Optional[str] = None
    source_url: Optional[str] = None
    is_custom: bool
    external_id: Optional[str] = None
    tags: Optional[str] = None
    calories: Optional[float] = None
    protein_g: Optional[float] = None
    carbs_g: Optional[float] = None
    fat_g: Optional[float] = None
    created_at: datetime
    updated_at: datetime
    ingredients: list[IngredientInRecipeResponse] = []

    model_config = {"from_attributes": True}


class PaginatedRecipes(BaseModel):
    items: list[RecipeListItem]
    total: int
    page: int
    per_page: int
    pages: int


class IngredientCategoryUpdate(BaseModel):
    category: str
